File: advanced_localization_cpp/src/failsafe.cpp
#include "advanced_localization_cpp/failsafe.hpp"

namespace advanced_localization
{

FailsafeManager::FailsafeManager(int max_rejections, double max_position_trace_m2)
: max_rejections_(max_rejections),
  max_position_trace_m2_(max_position_trace_m2)
{
}

FailsafeDecision FailsafeManager::evaluate(const FilterHealth & health, const std::optional<UpdateResult> & latest_update)
{
  std::string requested_mode = mode_;
  std::string reason = "hold";
  FailsafeCode code = FailsafeCode::None;

  if (mode_ == "AUTO") {
    if (health.consecutive_lightglue_rejections >= max_rejections_) {
      requested_mode = "LOITER";
      reason = "lightglue_rejection_limit";
      code = FailsafeCode::Inconsistency;
    } else if (health.consecutive_vision_loss >= max_rejections_) {
      requested_mode = "LOITER";
      reason = "vision_loss_limit";
      code = FailsafeCode::VisionLoss;
    } else if (health.position_trace_xy_m2 > max_position_trace_m2_) {
      requested_mode = "LOITER";
      reason = "covariance_limit";
      code = FailsafeCode::HighCovariance;
    }
  } else if (mode_ == "LOITER") {
    if (latest_update.has_value() && latest_update->accepted && health.position_trace_xy_m2 <= max_position_trace_m2_) {
      requested_mode = "AUTO";
      reason = "vision_recovered";
      code = FailsafeCode::None;
    }
  }

  const bool changed = requested_mode != mode_;
  mode_ = requested_mode;
  return FailsafeDecision{mode_, reason, code, changed};
}

}  // namespace advanced_localization

File: src/advanced_localization/failsafe.py
from __future__ import annotations

from dataclasses import dataclass

from .eskf import FilterHealth
from .types import FailsafeCode, UpdateResult


@dataclass(frozen=True)
class FailsafeDecision:
    mode: str
    reason: str
    code: FailsafeCode
    changed: bool


class FailsafeManager:
    def __init__(self, max_rejections: int, max_position_trace_m2: float) -> None:
        self._max_rejections = max_rejections
        self._max_position_trace_m2 = max_position_trace_m2
        self._mode = "AUTO"

    @property
    def mode(self) -> str:
        return self._mode

    def evaluate(self, health: FilterHealth, latest_update: UpdateResult | None) -> FailsafeDecision:
        requested_mode = self._mode
        reason = "hold"
        code = FailsafeCode.NONE

        if self._mode == "AUTO":
            if health.consecutive_lightglue_rejections >= self._max_rejections:
                requested_mode = "LOITER"
                reason = "lightglue_rejection_limit"
                code = FailsafeCode.INCONSISTENCY
            elif health.consecutive_vision_loss >= self._max_rejections:
                requested_mode = "LOITER"
                reason = "vision_loss_limit"
                code = FailsafeCode.VISION_LOSS
            elif health.position_trace_xy_m2 > self._max_position_trace_m2:
                requested_mode = "LOITER"
                reason = "covariance_limit"
                code = FailsafeCode.HIGH_COVARIANCE

        elif self._mode == "LOITER":
            if latest_update is not None and latest_update.accepted and health.position_trace_xy_m2 <= self._max_position_trace_m2:
                requested_mode = "AUTO"
                reason = "vision_recovered"
                code = FailsafeCode.NONE

        changed = requested_mode != self._mode
        if changed:
            self._mode = requested_mode

        return FailsafeDecision(mode=self._mode, reason=reason, code=code, changed=changed)
